angr/exploration_techniques/tracer.py

```diff
@@ -0,0 +1,291 @@
+import angr
+import claripy
+import logging
+
+from . import ExplorationTechnique
+from .oppologist import Oppologist
+
+from .. import SIM_LIBRARIES, BP_AFTER, BP_BEFORE
+
+from ..calling_conventions import SYSCALL_CC
+from ..errors import AngrTracerError
+from ..misc.tracer.simprocedures import receive
+
+l = logging.getLogger("angr.exploration_techniques.tracer")
+
+# global writable attribute used for specifying cache procedures
+#GlobalCacheManager = None
+
+class Tracer(ExplorationTechnique):
+    """
+    An exploration technique that follows an angr path with a concrete input.
+    The tracing result is the state after executing the last basic block of the
+    given trace and can be found in 'traced' stash.
+    
+    If the given concrete input makes the program crash, the last correct
+    states that you might want are kept in the 'predecessors' list. The crashed
+    state can be found with CrashMonitor exploration technique.
+    """
+
+    def __init__(self, trace=None, resiliency=True, dump_syscall=False, keep_predecessors=1):
+        """
+        :param trace            : The basic block trace.
+        :param resiliency       : Should we continue to step forward even if qemu and angr disagree?
+        :param dump_syscall     : True if we want to dump the syscall information.
+        :param keep_predecessors: Number of states before the final state we should preserve.
+                                  Default 1, must be greater than 0.
+        """
+
+        self._trace = trace
+        self._resiliency = resiliency
+        self._dump_syscall = dump_syscall
+
+        # keep track of the last basic block we hit
+        if keep_predecessors < 1:
+            raise ValueError("Must have keep_predecessors >= 1")
+        self._predecessors = [None] * keep_predecessors
+
+       #cm = LocalCacheManager(dump_cache=dump_cache) if GlobalCacheManager is None else GlobalCacheManager
+       ## cache managers need the tracer to be set for them
+       #self._cache_manager = cm
+       #self._cache_manager.set_tracer(self)
+
+       ## set by a cache manager
+       #self._loaded_from_cache = False
+
+       ## set up cache hook
+       #receive.cache_hook = self._cache_manager.cacher
+
+        # whether we should follow the trace
+        self._no_follow = self._trace is None
+
+        # initilize the syscall statistics if the flag is on
+        if self._dump_syscall:
+            self._syscall = []
+
+    def setup(self, simgr):
+        self.project = simgr._project
+        s = simgr.active[0]
+
+        # initialize the basic block counter to 0
+        s.globals['bb_cnt'] = 0
+
+        if self.project.loader.main_object.os == 'cgc':
```

Comment:
Why is syscall dumping only supported for CGC? I understand there's an explicit check for CGC terminate, but we can check for `os == 'cgc'` there instead of here.